2021/new_data_augustin/read_airnow.py

```diff
@@ -55,80 +57,116 @@
-    data.set_index('datetime', inplace=True)
-    #drop yy/mm/dd and hh:mm columns
-    data.drop(columns=['mm/dd/yy','hh:mm'], inplace=True)
-    
-    
-    
+    data = _calc_datetime(data)
+
     #list of available variables
     av_data = ['concbc', 'concco', 'concnh3', 'concno', 'concno2', 'concnox', 'concnoy', 'conco3', 'concpm10', 'concpm25', 'concso2']
     if vars_to_retrieve == None:
         vars_to_retrieve = av_data
-    
-    
+
     #convert dataframes to dictionnaries
-    dic_cfg = dict()
-    for column in cfg.columns:
-        dic_cfg[column] = np.array(cfg[column].values)
-    dic_data = dict()
-    for column in data.columns:
-        dic_data[column] = np.array(data[column].values)
+    dic_cfg, dic_data = _data_to_dicts(data, cfg)
 
-    
-    # list of stationData objects
-    print('create stationData objects')
-    stationsData = []
+    # list of stat objects
+    print('create stat objects')
+    stats = []
     for i in tqdm(range(len(cfg['aqsid']))):
         for var in vars_to_retrieve:
             try:
-                #initialize stationData object
-                stationData = pya.StationData()
-
-                # fill stationData with cfg
-                stationData['data_id'] = 'CAMS84_AIRNOW'
-                stationData['station_name'] = dic_cfg['name'][i]
-                stationData['station_id'] = dic_cfg['aqsid'][i]
-                stationData['latitude'] = dic_cfg['lat'][i]
-                stationData['longitude'] = dic_cfg['lon'][i]
-                stationData['altitude'] = dic_cfg['elevation'][i]
-                stationData['ts_type'] = 'hourly'
-
-                # fill stationData with data
-                mask = (dic_data['variable'] == pyvars[var]['var']) & (dic_data['station_id'] == stationData['station_id'])
-                stationData[var] = dic_data['value'][mask].astype('datetime64[s]')
-
-                # for each variable, there needs to be an entry in the var_info dict
-                stationData['var_info'][var] = dict()
-                stationData['var_info'][var]['units'] = pyvars[var]['unit']
-
-                stationsData.append(stationData)
+                stat = _make_station_data(dic_cfg, dic_data, i, var)
+
+                stats.append(stat)
             except KeyError:
                 print("Available variables: ",av_data)
                 raise
-    return stationsData
+    return stats
+
+if __name__ == '__main__':
+    path_data = '/lustre/storeA/project/aerocom/aerocom1/AEROCOM_OBSDATA/MACC_INSITU_AirNow'
+    dirs = [os.path.join(path_data, o) for o in os.listdir(path_data) if os.path.isdir(os.path.join(path_data,o))]
+    files = [glob(os.path.join(path_data, d, '*.dat')) for d in dirs][0]
```

Comment:
@AugustinMortier you are only picking the first file here from each sub-directory, right?

Reply:
There is in total more than 7000 *.dat files ... oh my, this will be terribly slow to read

Reply:
No, it's just that it's an array of array.. so the [0] allows to get rid of one dimension. There is probably a way to avoid that.

Reply:
No, that way you only take the files from the first directory (it is correct that `files` contains a list of lists, which contain the .dat files for each subdirectory, however, by using `[0]`, you only take the first list, which corresponds to the files in the first sub directory). An easy way to get all data files in all sub-dirs with a one-liner is 
```python
glob(f'{path_data}/**/*.dat')
```

I fixed this in the current version, which makes more than 7000 files to read in total.